### ISSUE TITLE
numpy.array gives unexpected results for generator expressions

### ISSUE THREAD

**Issue by arthur-tacca** [numpy/numpy#10144]
One might reasonably expect that `numpy.array` works equally well on generators as on lists (or range objects), if perhaps a little slower. In actual fact, the result is a zero-dimensional array containing the generator as an object:

    In[2]: import numpy as np
    In[3]: np.array([0, 1, 2])
    Out[3]: array([0, 1, 2])
    In[4]: np.array(range(3))
    Out[4]: array([0, 1, 2])
    In[5]: np.array(x for x in range(3))
    Out[5]: array(<generator object <genexpr> at 0x7f3e59f69db0>, dtype=object)

This has bitten me a couple of times, and each time it has become obvious what has happened fairly quickly, but has still been annoying enough that I am filing this report :-)

It would be nice if this worked correctly (like `fromiter`, which I have just learned about, except not needing a `dtype` specified and working recursively). But I think just an exception or a warning to the console would do. In my case, and I suspect many others, just turning the generator into a list comprehension was acceptable, so the key thing was just finding the source of the problem.

**Comment by seberg** [comment]
One problem is that generators are not sequences, you do not know their length and you cannot iterate through them twice. So internally, the only method is to convert every generator into a list first, and I guess nobody really could decide on actually doing that.
(some things like `range` are somewhat generators, but also sequences and are thus not problematic)

**Comment by ahaldane** [comment]
This one probably boils down to the `object` type being too greedy. We've discussed this quite extensively in the past (see #5353), and the plan is to make object type less greedy in the future by only casting to object if `dtype='O'` was explicitly supplied, but we haven't got around to it yet.

**Comment by seberg** [comment]
True, once this is/would be an error without `dtype=object`, making it work at least for any other dtype is pretty clearly nice. For object dtype itself, someone has to figure out how bad the compatibility break is.
